lib/membrane_rtc_engine/engine.ex

```diff
@@ -842,115 +805,105 @@ defmodule Membrane.RTC.Engine do
+        Membrane.Tee.PushOutput
+      end
 
-        _other_child, old_values ->
-          old_values
+    links = [
+      link({:endpoint, endpoint_id})
+      |> via_out(Pad.ref(:output, track_id))
+      |> to({:tee, track_id}, tee)
+    ]
+
+    endpoints_to_link =
+      ctx.children
+      |> Enum.flat_map(fn
+        {{:endpoint, endpoint_id}, _data} -> [endpoint_id]
+        _child -> []
       end)
+      |> Enum.filter(&MapSet.member?(state.waiting_for_linking[&1], track_id))
 
-    {waiting_for_linking, parent_spec}
-  end
+    waiting_for_linking =
+      Enum.reduce(endpoints_to_link, state.waiting_for_linking, fn endpoint_id,
+                                                                   waiting_for_linking ->
+        Map.update!(waiting_for_linking, endpoint_id, &MapSet.delete(&1, track_id))
+      end)
 
-  defp link_outbound_tracks(tracks, endpoint_id, ctx, state) do
-    Enum.reduce(tracks, {MapSet.new(), %ParentSpec{children: %{}, links: []}}, fn
-      {track_id, format}, {not_linked, parent_spec} ->
-        endpoint_tee = find_child(ctx, pattern: {:tee, {_other_endpoint_id, ^track_id}})
-        filter_tee = find_child(ctx, pattern: {:filter_tee, {_other_endpoint_id, ^track_id}})
-
-        {format_specific_tee, new_children, filter_links} =
-          cond do
-            format == :raw and filter_tee != nil ->
-              {filter_tee, %{}, []}
-
-            format == :raw and filter_tee == nil and endpoint_tee != nil ->
-              {:tee, endpoint_track_ids} = endpoint_tee
-              prepare_filter_tee(track_id, endpoint_track_ids, state)
-
-            true ->
-              {endpoint_tee, %{}, []}
-          end
-
-        if format_specific_tee do
-          endpoint_name = {:endpoint, endpoint_id}
-
-          new_link =
-            link(format_specific_tee)
-            |> via_out(Pad.ref(:output, endpoint_name))
-            |> via_in(Pad.ref(:input, track_id))
-            |> to({:endpoint, endpoint_id})
-
-          {not_linked,
-           %ParentSpec{
-             children: Map.merge(parent_spec.children, new_children),
-             links: parent_spec.links ++ filter_links ++ [new_link]
-           }}
+    {endpoints_handling_raw_format, endpoints_handling_remote_format} =
+      Enum.split_with(endpoints_to_link, fn endpoint_id ->
+        format = state.subscriptions[endpoint_id][track_id]
+        format == :raw
+      end)
+
+    links =
+      links ++
+        if endpoints_handling_raw_format == [] do
+          []
         else
-          {MapSet.put(not_linked, track_id), parent_spec}
-        end
+          prepare_raw_format_links(track_id, state)
+        end ++
+        Enum.map(endpoints_handling_raw_format, fn endpoint_id ->
+          prepare_track_to_endpoint_links(track_id, endpoint_id, :raw_format_tee)
+        end) ++
+        Enum.map(endpoints_handling_remote_format, fn endpoint_id ->
+          prepare_track_to_endpoint_links(track_id, endpoint_id, :tee)
+        end)
```

Comment:
IMO this fragment is too complex and it is hard to understand what we will be in the result list. What do you think about:
1. Evaluating these maps and if before concatenating lists?
2. As both maps look similar, could we create a function like `map_endpoints_to_links(endpoints, track_id, formats)` and move mapping to this function. Then concatenation would look like this:
```elixir
        links ++
        if endpoints_handling_raw_format == [] do
          []
        else
          prepare_raw_format_links(track_id, state)
        end ++
        map_endpoints_to_links(endpoints_handling_raw_format, track_id, :raw_format_tee) ++
        map_endpoints_to_links(endpoints_handling_remote_format, track_id, :tee)
```